### PR TITLE
follow-user

### PR DESCRIPTION
**Pull request by lannp**

### CODE REVIEW

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+
+  def create
+    @user = User.find params[:followed_id]
+    current_user.follow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
```
Comment (hienbx94) [inline app/controllers/relationships_controller.rb]: indent

File: app/controllers/relationships_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      format.js
+    end
+  end
+
+  def destroy
+    @user = Relationship.find(params[:id]).followed
+    current_user.unfollow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
```
Comment (thiennv57) [inline app/controllers/relationships_controller.rb]: Ident

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ def create
     user = User.find_by email: params[:session][:email].downcase
     if user && user.authenticate(params[:session][:password])
       log_in user
-      params[:session][:remember_me] == Settings.remember ? remember(user) : forget(user)
+      params[:session][:remember_me] == Settings.remember ? remember(user): forget(user)
```
Comment (thiennv57) [inline app/controllers/sessions_controller.rb]: remove ()

File: db/seeds.rb [hunk 4]
```diff
@@ -10,3 +10,19 @@
...
+              email: email,
+              password:              password,
+              password_confirmation: password)
+end
+
+users = User.all
+user  = users.first
+following = users[2..50]
+followers = users[3..40]
+following.each { |followed| user.follow(followed) }
```
Comment (hienbx94) [inline db/seeds.rb]: indent

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -38,6 +39,19 @@ def update
     end
   end
 
+  def following
+    @title = t "user_follow.following"
+    @user  = User.find(params[:id])
```
Comment (ttuan) [inline app/controllers/users_controller.rb]: bỏ ngoặc

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -38,6 +39,19 @@ def update
...
+    @users = @user.following.paginate(page: params[:page])
```
Comment (thiennv57) [inline app/controllers/users_controller.rb]: remove ()

Comment (ttuan) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/models/user.rb [hunk 7]
```diff
@@ -35,9 +43,23 @@ def forget
...
+  def follow other_user
+    active_relationships.create(followed_id: other_user.id)
+  end
+
+  def unfollow other_user
+    active_relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  def following? other_user
+    following.include?(other_user)
```
Comment (thiennv57) [inline app/models/user.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -38,6 +39,19 @@ def update
...
+    render "show_follow"
+  end
+
+  def followers
+    @title = t "user_follow.followers"
+    @user  = User.find(params[:id])
```
Comment (ttuan) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,7 +1,8 @@
 class UsersController < ApplicationController
-  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :logged_in_user, only: [:index, :edit, :update,
+    :following, :followers]
   before_action :correct_user, only: [:edit, :update]
-  
+
   def show
     @user = User.find params[:id]
```
Comment (VuongLong) [inline app/controllers/users_controller.rb]: Nên dùng macro tạo 1 function private init_user sau đó before_action cho edit show ...

File: app/views/users/_unfollow.html.erb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
+  html: { method: :delete }) do |f| %>
```
Comment (thiennv57) [inline app/views/users/_unfollow.html.erb]: indent

File: app/views/static_pages/home.html.erb [hunk 11]
```diff
@@ -1,4 +1,13 @@
 <div class="center jumbotron">
+  <% if logged_in? %>
+    <div class="row">
+      <aside class="col-md-4">      
+        <section class="stats">
+          <%= render "shared/stats" %>
+        </section>
+    </div>
+  <% else %>
   <h1><%= t "static_pages.welcome" %></h1>
```
Comment (ttuan) [inline app/views/static_pages/home.html.erb]: indent

File: config/routes.rb [hunk 12]
```diff
@@ -15,6 +15,13 @@
   post   "login"   => "sessions#create"
   delete "logout"  => "sessions#destroy"
   resources :users
+  resources :users do
+    member do
+      get :following, :followers
+    end
+  end
+
+  resources :relationships,       only: [:create, :destroy]
```
Comment (thiennv57) [inline config/routes.rb]: indent

File: app/views/users/_follow_form.html.erb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<% unless current_user?(@user) %>
```
Comment (ttuan) [inline app/views/users/_follow_form.html.erb]: remove ()

File: app/views/users/_follow_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,5 @@
...
+  <div id="follow_form">
+  <%= render current_user.following?(@user) ? "unfollow" : "follow" %>
```
Comment (ttuan) [inline app/views/users/_follow_form.html.erb]: indent

Comment (ttuan) [inline app/views/users/_follow_form.html.erb]: remove ()

File: app/views/users/_unfollow.html.erb (hunk 10 above)
Comment (ttuan) [inline app/views/users/_unfollow.html.erb]: remove space after and before {}

File: app/views/shared/_stats.html.erb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <strong id="following" class="stat">
+    <%= @user.following.count %>
+  </strong>
+  <%= t"user_following.following" %>
```
Comment (ThuBM) [inline app/views/shared/_stats.html.erb]: add space after t

File: app/views/shared/_stats.html.erb [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+  <strong id="followers" class="stat">
+    <%= @user.followers.count %>
+  </strong>
+  <%= t"user_following.followers" %>
```
Comment (ThuBM) [inline app/views/shared/_stats.html.erb]: add space after t

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,9 +1,10 @@
 class UsersController < ApplicationController
-  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :logged_in_user, only: [:index, :edit, :update,
+    :following, :followers]
```
Comment (trantube94) [inline app/controllers/users_controller.rb]: remove following,followers

File: app/views/static_pages/home.html.erb [hunk 18]
```diff
@@ -1,4 +1,14 @@
...
+  <% if logged_in? %>
+    <div class="row">
+      <aside class="col-md-4">      
+        <section class="stats">
+          <%= render "shared/stats" %>
+        </section>
+      </aside>
+    </div>
+  <% else %>
   <h1><%= t "static_pages.welcome" %></h1>
```
Comment (ThuBM) [inline app/views/static_pages/home.html.erb]: indent

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -15,7 +16,6 @@ def new
   end
 
   def edit
-    @user = User.find params[:id]
   end
```
Comment (trantube94) [inline app/controllers/users_controller.rb]: remove def edit

File: app/views/shared/_stats.html.erb (hunk 15 above)
Comment (trantube94) [inline app/views/shared/_stats.html.erb]: space after t

**Comment by ThuBM** [comment]
OK

**Pull request merged**